2020-02-Spring/3_trees/README.md

```diff
@@ -141,21 +211,287 @@ Go to [Top](#top)
+    * O(N) - In the worst case, every node is visited again while searching for an inorder predecessor, in the case of a degenerate tree
+  
+* ***Space Complexity***: `O(1)`
+
+    * No new data structures are created. Only pointers are used.
+
+
+<details>
+<summary>Click to see Morris traversal solution</summary>
+
+```java
+
+// Helper method for swapping the values of two nodes
+public void swap(TreeNode a, TreeNode b) {
+  int temp = a.val;
+  a.val = b.val;
+  b.val = temp;
+}
+
+public void recoverTree(TreeNode root) {
+  TreeNode first = null;
+  TreeNode second = null;
+  TreeNode predecessor = null; // For Morris traversal
+  TreeNode previous = null;    // For finding the swapped nodes
+
+  while (root != null) {
+    if (root.left != null) {
+      // Find inorder predecessor of root
+      predecessor = root.left;
+      while (predecessor.right != null && predecessor.right != root) {
+        predecessor = predecessor.right;
+      }
+      // If there is no link to the parent node
+      // Create one, then visit the left subtree
+      if (predecessor.right == null) {
+        predecessor.right = root;
+        root = root.left;
+      } else {
+        // Link already exists. We're done with the left subtree
+
+        // Check to see if this node is swapped
+        if (previous != null && root.val < previous.val) {
+          second = root;
+          if (first == null) first = previous;
+        }
+        previous = root;          // Save this node for swap comparison
+
+        predecessor.right = null; // Break the existing link
+        root = root.right;        // Go to the right subtree
+      }
+    } else {
+      // No left subtree. No need to check for predecessors. Just go right.
+      
+      // Check to see if this node is swapped
+      if (previous != null && root.val < previous.val) {
+        second = root;
+         if (first == null) first = previous;
+      }
+      previous = root;          // Save this node for swap comparison
+      
+      root = root.right; // Go to the right subtree
+    }
+  }
+  // We are done traversing through the tree.
+  // We can swap the nodes now.
+  swap(first, second);
+}
+
+```
+
+</details>
+
+</details>
+
+
+#### Driver for Solution
+
+See [the full solution file](./recover_bst/java/Solution.java) for helper methods.
+
+```java
+
+    public static void main(String[] args) {
+        List<Integer> treeList1 = new ArrayList<>(Arrays.asList(1, 3, null, null, 2));
+        List<Integer> treeList2 = new ArrayList<>(Arrays.asList(3, 1, 4, null, null, 2));
+
+        TreeNode tree1 = buildTree(treeList1);
+        TreeNode tree2 = buildTree(treeList2);
+
+        printTree(tree1); // Before: [ 3 2 1 ]
+        printTree(tree2); // Before: [ 1 3 2 4 ]
+
+        recoverTree(tree1);
+        recoverTree(tree2);
+
+        printTree(tree1); // After: [ 1 2 3 ]
+        printTree(tree2); // After: [ 1 2 3 4 ]
+    }
+
+```
 
```

Comment:
Including helper methods like printTree & buildTree would be helpful for the participants to use this driver here. Add a drop-down if this section becomes too long.

Reply:
If those methods are added, they shouldn't be in the README.

Reply:
I think I'm going to leave them out since there's a link to the solution file in the README.md already